crates/gen/src/method.rs

```diff
@@ -135,116 +137,101 @@ impl Method {
-            }
-        }
-    }
+        let vtable_offset = Literal::u32_unsuffixed(self.vtable_offset);
 
-    pub fn gen_composable(&self, interface: &RequiredInterface) -> TokenStream {
-        let method_name = format_ident(&self.name);
-        let interface = format_ident(&interface.name.name);
+        let vcall = if let Some(return_type) = &self.return_type {
+            let return_arg = return_type.gen_abi_return_arg();
 
-        if self.params.len() == 2 {
-            // For non-aggregation scenarios this is just a default constructor, hence the
-            // "new" method name in line with non-composable default constructors.
-            quote! {
-                pub fn new() -> ::winrt::Result<Self> {
-                    Self::#interface(|f| f.#method_name(::std::option::Option::None, &mut ::std::option::Option::None))
+            if return_type.array {
+                quote! {
+                    let mut result__: #return_type_tokens = ::std::mem::zeroed();
+                    (::winrt::Interface::vtable(this).#vtable_offset)(::winrt::Abi::abi(this), #(#args)* #composable_args #return_arg)
+                        .and_then(|| result__ )
+                }
+            } else {
+                quote! {
+                    let mut result__: <#return_type_tokens as ::winrt::Abi>::Abi = ::std::mem::zeroed();
+                        (::winrt::Interface::vtable(this).#vtable_offset)(::winrt::Abi::abi(this), #(#args)* #composable_args #return_arg)
+                            .from_abi::<#return_type_tokens>(result__ )
                 }
             }
         } else {
-            let params = &self.params[..self.params.len() - 2];
-            let constraints = gen_constraint(params);
-            let args = gen_arg(params);
-            let params = gen_param(params);
-
             quote! {
-                pub fn #method_name<#constraints>(#params) -> ::winrt::Result<Self> {
-                    Self::#interface(|f| f.#method_name(#args ::std::option::Option::None, &mut ::std::option::Option::None))
+                (::winrt::Interface::vtable(this).#vtable_offset)(::winrt::Abi::abi(this), #(#args)* #composable_args).ok()
+            }
+        };
+
+        match kind {
+            InterfaceKind::Default => quote! {
+                pub fn #method_name<#constraints>(&self, #params) -> ::winrt::Result<#return_type_tokens> {
+                    unsafe {
+                        let this = self;
+                        #vcall
+                    }
+                }
+            },
+            InterfaceKind::NonDefault | InterfaceKind::Overrides => {
+                let interface = interface.gen();
+                quote! {
+                    pub fn #method_name<#constraints>(&self, #params) -> ::winrt::Result<#return_type_tokens> {
+                        unsafe {
+                            let this = &::winrt::Interface::cast::<#interface>(self).unwrap();
```

Comment:
nit: I believe this call is safe and so it doesn't need to be in the unsafe block.